### DISCUSSION TITLE
Schema transformations

### DISCUSSION THREAD

**Discussion by matixezor** [graphql-java/graphql-java#2504] [Q&A] [answered]
Hi,
I am trying to transform the schema in `instrumentation` using `SchemaTransformer`. The problem is that I would like to transform the schema only for particular requests(using some Http Headers to filter them). It seems that changes made by schemaTransformer are permanent. Is it possible to change the schema only in the context of a particular request?
Was trying to use `schema.transform()` and it did work the way i described above. The problem with it was just getting to those nested fields was kinda complicated.

For example in `instrumentSchema` method, based on a condition(http Header) I can transform the schema where I change a fields type from `Email` to `String`:
```java   
if (headerFlag) {
     ...
     newSchema = schema.transform(
         b -> {
            b.clearAdditionalTypes();
            b.query(newQuery);
          }
      );
}

return newSchema;
```
So that I can send a query without header:
```
query{
    ...
    email{   // -> Email
      domain
    }
 
}
``` 
or with a header 
```
query{
    ...
    email   // -> String
}
``` 

Now I am trying do to it with:
```java
  private GraphQLSchema transformSchema(GraphQLSchema schema, GraphQLOutputType type, String fieldName) {
    return schemaTransformer.transform(schema, new GraphQLTypeVisitorStub() {
      @Override
      public TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context) {
        if (fieldDefinition.getName().equals(fieldName)) {
          GraphQLFieldDefinition changedNode = fieldDefinition.transform(builder -> builder.type(type));
          return changeNode(context, changedNode);
        }
        return TraversalControl.CONTINUE;
      }
    });
  }
  ```
But once I transform the schema(its copy to be precise) this way, the change is permanent and I can't ask in the next request for ` email { domain } ` as I am getting 
`Validation error of type SubSelectionNotAllowed: Sub selection not allowed on leaf type String of field email`.
Turns out even though I am transforming the copy of the original schema, the original schema gets transformed/overwritten as well.
Is that intended behaviour?

**Comment by bbakerman** [comment]
> Turns out even though I am transforming the copy of the original schema, the original schema gets transformed/overwritten as well.

Not quite - schemas are immutable and hence a "new" schema object is created

The changed schema (from the instrumentation) should be passed down after that and available to later steps.  

The changed version is never assigned back so I cant see how it can "overwrite any thing"

```
            InstrumentationExecutionParameters instrumentationParameters = new InstrumentationExecutionParameters(executionInput, this.graphQLSchema, instrumentationState);
                        InstrumentationContext<ExecutionResult> executionInstrumentation = instrumentation.beginExecution(instrumentationParameters);


            GraphQLSchema graphQLSchema = instrumentation.instrumentSchema(this.graphQLSchema, instrumentationParameters);
```

(on a side note I can see some edge cases where we should do `instrumentSchema` before ` beginExecution` but I digress)

If you can I would "transform" the schema BEFORE you create the GraphQL object per request.  Its cleaner than doing it inside the `GraphQL.execute` call via Instrumentation.

```
   GraphqlSchema newSchema  = tranformSchema(startingSchema, requestParamsForDecisions)
   GraphQl graphQL = GraphQl.newGraphQL(newSchema).build()
```

Transforming schemas can be slow if you have really large schemas - but creating GraphQl objects is super cheap.

**Reply by matixezor** [reply]
After some digging, it turns out that in our case in the schema the`GraphQLFieldDefinition original type` is a `GraphQLTypeReference`.
The original schema and newSchema hold the same type reference, that's why I guess the change is visible in both schemas

**Reply by matixezor** [reply]
Original schema before before calling schemaTransformer
<img width="813" alt="Zrzut ekranu 2021-08-6 o 19 50 51" src="https://user-images.githubusercontent.com/42327944/128552011-4280e8a1-5839-4dd4-8947-52c9046621f3.png">
New schema returned by schemaTransformer
<img width="807" alt="Zrzut ekranu 2021-08-6 o 19 49 21" src="https://user-images.githubusercontent.com/42327944/128552027-72b03816-9135-497e-ba81-aaf4970cece0.png">
Original schema after schemaTransformer execution
<img width="844" alt="Zrzut ekranu 2021-08-6 o 19 45 11" src="https://user-images.githubusercontent.com/42327944/128552034-c07c0558-ff53-4b4b-b8a3-b80a076e6e89.png">

**Reply by bbakerman** [reply]
So are you saying that this problem is one where you mistakenly thought something or you have found a bug

I am not quite sure from the above

**Reply by matixezor** [reply]
I found a bug. In our project we are also using graphql kickstart libs.
Using code below you can recreate it yourself:

```java
import graphql.kickstart.tools.GraphQLQueryResolver
import graphql.kickstart.tools.SchemaParser
import graphql.schema.DataFetchingEnvironment
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLSchema
import graphql.schema.GraphQLSchemaElement
import graphql.schema.GraphQLTypeVisitorStub
import graphql.schema.SchemaTransformer
import graphql.util.TraversalControl
import graphql.util.TraverserContext
import spock.lang.Specification

import static graphql.util.TreeTransformerUtil.changeNode

class SchemaTransformationTest extends Specification{

    class TestUser {
        private String id

        TestUser(String id) {
            this.id = id
        }
    }

    class TestUserQueryResolver implements GraphQLQueryResolver {
         TestUser getTestUser(
                String id, DataFetchingEnvironment dfe
        ) {
            return TestUser(id)
        }
    }

    def schemaTransformer = new SchemaTransformer()

    def transformSchema(schema) {
        return schemaTransformer.transform(schema, new GraphQLTypeVisitorStub() {

            @Override
            TraversalControl visitGraphQLFieldDefinition(GraphQLFieldDefinition fieldDefinition, TraverserContext<GraphQLSchemaElement> context) {
                if (fieldDefinition.name == "id") {
                    def changedNode = fieldDefinition.transform({ builder -> builder.name("idChanged") })
                    return changeNode(context, changedNode)
                }
                return TraversalControl.CONTINUE
            }
        })
    }

    def schemaString = """
                        type TestUser {
                            id: String
                        }
                        type Query {
                            testUser(id: String): TestUser
                        }
                    """

    def "can change field in schema created using graphql-tools parser"() {
        given:
        def schema = SchemaParser
                .newParser()
                .schemaString(schemaString)
                .resolvers(new TestUserQueryResolver())
                .build()
                .makeExecutableSchema()

        when:
        GraphQLSchema newSchema = transformSchema(schema)

        then:
        newSchema != schema
        newSchema.getQueryType().getFieldDefinition("testUser").getType() != schema.getQueryType().getFieldDefinition("testUser").getType()
        newSchema.getQueryType().getFieldDefinition("testUser").getType().getChildren().get(0) != schema.getQueryType().getFieldDefinition("testUser").getType().getChildren().get(0)
    }

    def "can change field in schema created using graphql-java parser"() {
        given:
        def schema = TestUtil.schema(schemaString) // TestUtil class from your graphql-java tests

        when:
        GraphQLSchema newSchema = transformSchema(schema)

        then:
        newSchema != schema
        newSchema.getQueryType().getFieldDefinition("testUser").getType() != schema.getQueryType().getFieldDefinition("testUser").getType()
        newSchema.getQueryType().getFieldDefinition("testUser").getType().getChildren().get(0) != schema.getQueryType().getFieldDefinition("testUser").getType().getChildren().get(0)
    }
}

```

**Reply by matixezor** [reply]
@bbakerman do you have an advice on working this out?

**Comment by bbakerman** [comment]
I cant debug kickstart code - it is not part of this project at all.

**Comment by matixezor** [comment, marked as answer]
To anyone that comes with similar problem - adding such a method to visitor solved the problem. It ensures that new schema has new references.
```kotlin
override fun visitGraphQLTypeReference(
    node: GraphQLTypeReference,
    context: TraverserContext<GraphQLSchemaElement>
): TraversalControl =
    TreeTransformerUtil.changeNode(context, node)
```
